cmd/ci-secret-bootstrap/main.go

```diff
@@ -0,0 +1,247 @@
+		return fmt.Errorf("the mandatory arg bw-password-path is empty")
+	}
+	if o.configPath == "" {
+		return fmt.Errorf("the mandatory arg config is empty")
+	}
+	return nil
+}
+
+func (o *options) completeOptions() error {
+	bytes, err := ioutil.ReadFile(o.bwPasswordPath)
+	if err != nil {
+		return err
+	}
+	o.bwPassword = strings.TrimSpace(string(bytes))
+
+	kubeConfigs, _, err := util.LoadKubeConfigs(o.kubeConfigPath)
+	if err != nil {
+		return err
+	}
+	o.kubeConfigs = kubeConfigs
+
+	config, err := loadConfig(o.configPath)
+	if err != nil {
+		return err
+	}
+	o.config = config
+
+	return o.validateCompletedOptions()
+}
+
+func (o *options) validateCompletedOptions() error {
+	if o.bwPassword == "" {
+		return fmt.Errorf("found empty file: '%s'", o.bwPasswordPath)
+	}
+	if len(o.config) == 0 {
+		return fmt.Errorf("found empty config")
+	}
+	toSet := sets.NewString()
+	for _, secretConfig := range o.config {
+		if len(secretConfig.From) == 0 {
+			return fmt.Errorf("found empty 'From'")
+		}
+		if len(secretConfig.To) == 0 {
+			return fmt.Errorf("found empty 'To'")
+		}
+		for key, bwContext := range secretConfig.From {
+			if key == "" {
+				return fmt.Errorf("found empty key whose value is '%v'", bwContext)
+			}
+			if bwContext.BWItem == "" {
+				return fmt.Errorf("found empty bw_item value for key '%s'", key)
+			}
+			if bwContext.Field == "" && bwContext.Attachment == "" {
+				return fmt.Errorf("both field and attachment are empty for key '%s'", key)
+			}
+			if bwContext.Field != "" && bwContext.Attachment != "" {
+				return fmt.Errorf("cannot use both field and attachment are for key '%s'", key)
+			}
+		}
+		for _, secretContext := range secretConfig.To {
+			if secretContext.Cluster == "" {
+				return fmt.Errorf("found empty cluster value: '%v'", secretContext)
+			}
+			if secretContext.Namespace == "" {
+				return fmt.Errorf("found empty namespace value: '%v'", secretContext)
+			}
+			if secretContext.Name == "" {
+				return fmt.Errorf("found empty name value: '%v'", secretContext)
+			}
+			if _, ok := o.kubeConfigs[secretContext.Cluster]; !ok {
+				return fmt.Errorf("failed to find cluster context in the kubeconfig: '%s'", secretContext.Cluster)
+			}
+			secretContextString := fmt.Sprintf("%v", secretContext)
+			if toSet.Has(secretContextString) {
+				return fmt.Errorf("secret context '%s' showed more than once in the config", secretContextString)
+			}
+			toSet.Insert(secretContextString)
+		}
+	}
+	return nil
+}
+
+type Config = []SecretConfig
+
+type BWContext struct {
+	BWItem     string `json:"bw_item"`
+	Field      string `json:"field,omitempty"`
+	Attachment string `json:"attachment,omitempty"`
+}
+
+type SecretContext struct {
+	Cluster   string `json:"cluster"`
+	Namespace string `json:"namespace"`
+	Name      string `json:"name"`
+}
+
+type SecretConfig struct {
+	From map[string]BWContext `json:"from"`
+	To   []SecretContext      `json:"to"`
+}
+
+func loadConfig(path string) (Config, error) {
```

Comment:
Why is this a function? Is it called more than once?

Reply:
Removed.